lib/Dialect/FIRRTL/Transforms/RandomizeRegisterInit.cpp

```diff
@@ -0,0 +1,85 @@
+//===- RandomizeRegisterInit.cpp - Randomize register initialization ------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the RandomizeRegisterInit pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/AnnotationDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
+#include "circt/Dialect/FIRRTL/FIRRTLInstanceGraph.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "mlir/IR/Builders.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+struct RandomizeRegisterInitPass
+    : public RandomizeRegisterInitBase<RandomizeRegisterInitPass> {
+  void runOnOperation() override;
+};
+
+} // end anonymous namespace
+
+std::unique_ptr<mlir::Pass> circt::firrtl::createRandomizeRegisterInitPass() {
+  return std::make_unique<RandomizeRegisterInitPass>();
+}
+
+void RandomizeRegisterInitPass::runOnOperation() {
+  FModuleOp op = getOperation();
+  CircuitOp circuit = op->getParentOfType<CircuitOp>();
+  OpBuilder builder(op);
+
+  // Look for a DUT annotation.
+  FModuleOp dut;
+  for (auto mod : circuit.getOps<FModuleOp>()) {
+    if (AnnotationSet(mod).hasAnnotation(dutAnnoClass)) {
+      dut = mod;
+      break;
+    }
+  }
+
+  // If there is a DUT, and this module is not a child of it, return early.
+  if (dut) {
+    auto instanceGraph = InstanceGraph(circuit);
```

Comment:
I think `InstanceGraph(circuit)` will walk every operation in the circuit so we should avoid this. I guess it is difficult to query `getAnaysis<InstanceGraph>` from a module pass, so we might have to convert  `RandomizeRegisterInitPass` into circuit op pass, and then internally parallelize it.

Reply:
👍  yeah, i will update this to a circuit pass and internally parallelize the module visits.